nexus/failure.go

```diff
@@ -0,0 +1,239 @@
+package nexus
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// A Failure represents failed handler invocations as well as `failed` or `canceled` operation results. Failures
+// shouldn't typically be constructed directly. The SDK APIs take a [FailureConverter] instance that can translate
+// language errors to and from [Failure] instances.
+type Failure struct {
+	// A simple text message.
+	Message string `json:"message,omitempty"`
+	// Stack trace which may be set if this error was generated by a language that supports it.
+	StackTrace string `json:"stackTrace,omitempty"`
+	// A key-value mapping for additional context. Useful for decoding the 'details' field, if needed.
+	Metadata map[string]string `json:"metadata,omitempty"`
+	// Additional JSON serializable structured data.
+	Details json.RawMessage `json:"details,omitempty"`
+	// An optional nested cause.
+	Cause *Failure `json:"cause,omitempty"`
+}
+
+// An error that directly represents a wire representation of [Failure].
+// The SDK will convert to this error by default unless the [FailureConverter] instance is customized.
+type FailureError struct {
+	// The underlying Failure object this error represents.
+	Failure Failure
+	// An optional nested cause.
+	Cause error
+}
+
+// Error implements the error interface.
+func (e *FailureError) Error() string {
+	return e.Failure.Message
+}
+
+// Unwrap implements the error interface.
+func (e *FailureError) Unwrap() error {
+	return e.Cause
+}
+
+// FailureConverter is used by the framework to transform [error] instances to and from [Failure] instances.
+// To customize conversion logic, implement this interface and provide your implementation to framework methods such as
+// [NewHTTPClient] and [NewHTTPHandler].
+// By default the SDK translates [HandlerError], [OperationError] and [FailureError] to and from [Failure]
+// objects maintaining their cause chain.
+// Arbitrary errors are translated to a [Failure] object with its Message set to the Error() string, losing the cause
+// chain.
```

Comment:
Might as well mark it as experimental since it's not exposed to Temporal applications yet.

```suggestion
// chain.
//
// NOTE: Experimental
```